[Windows][gfx110x-all] `hipthreads_examples/saxpy` is failing only on `CS-RORDMZ` runner group

### Description

`hipthreads` examples and tests were merged last week in https://github.com/ROCm/TheRock/pull/6770. Since then, it has been flaky on Windows `gfx110x-all`.

When hipthreads examples are run on the `azure-windows-11-gfx1101-gpu-rocm-runner` group, it passes (assuming runner is healthy, no driver issues).

When run on the `CS-RORDMZ` group, it always fails with exit code `3221225477`.
https://github.com/ROCm/TheRock/actions/runs/31002764024/job/92398573953?pr=7122#step:13:71
```
INFO: ++ Build [saxpy]$ cmake --build 'C:\actions-runner\_work\TheRock\TheRock\build\hipthreads\examples\saxpy\step3-simdize\build'
[1/2] Building HIP object CMakeFiles/saxpy.dir/main.hip.obj
[2/2] Linking HIP executable bin\saxpy.exe
clang++: warning: argument unused during compilation: '--rtlib=compiler-rt' [-Wunused-command-line-argument]
INFO: ++ Run [saxpy]$ 'C:\actions-runner\_work\TheRock\TheRock\build\hipthreads\examples\saxpy\step3-simdize\build\bin\saxpy.exe'
ERROR: saxpy exited with code 3221225477
ERROR: ---- last output ----
Exception Code: 0xC0000005
0x00007FFE607A5007, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x465007 byte(s), hipHccModuleLaunchKernel() + 0xC7E97 byte(s)
0x00007FFE607E01B4, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x4A01B4 byte(s), hipHccModuleLaunchKernel() + 0x103044 byte(s)
0x00007FFE60B9F878, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x85F878 byte(s), hipHccModuleLaunchKernel() + 0x4C2708 byte(s)
0x00007FFE60B9F933, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x85F933 byte(s), hipHccModuleLaunchKernel() + 0x4C27C3 byte(s)
0x00007FFE607E1F17, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x4A1F17 byte(s), hipHccModuleLaunchKernel() + 0x104DA7 byte(s)
0x00007FFE6071A9BC, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x3DA9BC byte(s), hipHccModuleLaunchKernel() + 0x3D84C byte(s)
0x00007FFE6070F270, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x3CF270 byte(s), hipHccModuleLaunchKernel() + 0x32100 byte(s)
0x00007FFE6072637B, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x3E637B byte(s), hipHccModuleLaunchKernel() + 0x4920B byte(s)
0x00007FFE603DCE2B, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x9CE2B byte(s), hipGetDeviceProperties() + 0x2624B byte(s)
0x00007FFE603EA173, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0xAA173 byte(s), hipGetDeviceProperties() + 0x33593 byte(s)
0x00007FFE60359020, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x19020 byte(s), hipGetCmdName() + 0x169D0 byte(s)
0x00007FFE60638BF5, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x2F8BF5 byte(s), hipRegisterTracerCallback() + 0x143775 byte(s)
0x00007FFE605F3A47, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x2B3A47 byte(s), hipRegisterTracerCallback() + 0xFE5C7 byte(s)
0x00007FFE605F1FA8, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x2B1FA8 byte(s), hipRegisterTracerCallback() + 0xFCB28 byte(s)
0x00007FFE606D7D00, C:\WINDOWS\SYSTEM32\amdhip64_7.dll(0x00007FFE60340000) + 0x397D00 byte(s), hipLaunchKernel() + 0x70 byte(s)
0x00007FF66FCC3938, C:\actions-runner\_work\TheRock\TheRock\build\hipthreads\examples\saxpy\step3-simdize\build\bin\saxpy.exe(0x00007FF66FCC0000) + 0x3938 byte(s)
0x00007FF66FCC3740, C:\actions-runner\_work\TheRock\TheRock\build\hipthreads\examples\saxpy\step3-simdize\build\bin\saxpy.exe(0x00007FF66FCC0000) + 0x3740 byte(s)
0x00007FF66FCC1307, C:\actions-runner\_work\TheRock\TheRock\build\hipthreads\examples\saxpy\step3-simdize\build\bin\saxpy.exe(0x00007FF66FCC0000) + 0x1307 byte(s)
0x00007FF66FCC9BF0, C:\actions-runner\_work\TheRock\TheRock\build\hipthreads\examples\saxpy\step3-simdize\build\bin\saxpy.exe(0x00007FF66FCC0000) + 0x9BF0 byte(s)
0x00007FFEBE31E8D7, C:\WINDOWS\System32\KERNEL32.DLL(0x00007FFEBE2F0000) + 0x2E8D7 byte(s), BaseThreadInitThunk() + 0x17 byte(s)
0x00007FFEBF42C40C, C:\WINDOWS\SYSTEM32\ntdll.dll(0x00007FFEBF3A0000) + 0x8C40C byte(s), RtlUserThreadStart() + 0x2C byte(s)

Traceback (most recent call last):
  File "C:\actions-runner\_work\TheRock\TheRock\build_tools\github_actions\test_executable_scripts\test_hipthreads_examples.py", line 280, in <module>
    run_example(example, binary, environ_vars)
  File "C:\actions-runner\_work\TheRock\TheRock\build_tools\github_actions\test_executable_scripts\test_hipthreads_examples.py", line 201, in run_example
    raise subprocess.CalledProcessError(result.returncode, cmd)
subprocess.CalledProcessError: Command '['C:\\actions-runner\\_work\\TheRock\\TheRock\\build\\hipthreads\\examples\\saxpy\\step3-simdize\\build\\bin\\saxpy.exe']' returned non-zero exit status 3221225477.
Error: Process completed with exit code 1.
```

### Bump PRs

Current bump PRs are exhibiting this same behaviour.
Run on azure and passed:
https://github.com/ROCm/TheRock/pull/7117
https://github.com/ROCm/TheRock/actions/runs/30990633880/job/92398540219?pr=7117

https://github.com/ROCm/TheRock/pull/7123
https://github.com/ROCm/TheRock/actions/runs/31006885617/job/92409829637?pr=7123

Run on CS-RORDMZ and failed:
https://github.com/ROCm/TheRock/pull/7122
https://github.com/ROCm/TheRock/actions/runs/31002764024/job/92398573953?pr=7122

### TheRock `main`

Nonexhaustive table of test results from commits on TheRock `main`:

| Commit | Runner group | Result |
|-|-|-|
| https://github.com/ROCm/TheRock/commit/a4dc044cb90580c6ff59eb591957dda0fa6d3ff9 (introduced hipthreads tests) | azure | [Passed ](https://github.com/ROCm/TheRock/actions/runs/30558341433/job/90958565337)|
https://github.com/ROCm/TheRock/commit/42dfe247f7a7a05964a6f31563f7cd348f9496c0 | CS-RORDMZ | [Failed](https://github.com/ROCm/TheRock/actions/runs/30614438499/job/91130413601) |
| https://github.com/ROCm/TheRock/commit/5a7aebd49035760358a9538b05b4299022a311e0 | CS-RORDMZ | [Failed](https://github.com/ROCm/TheRock/actions/runs/30641090441/job/91218988896) |
| https://github.com/ROCm/TheRock/commit/c5226c373b06938232391e5a67a38ff0acf8dd5a | azure | [Failed](https://github.com/ROCm/TheRock/actions/runs/30665102481/job/91294472528) (driver issue) |
| https://github.com/ROCm/TheRock/commit/b4cbb6a47b633f083f48a76620d9c8b847de147b | CS-RORDMZ | [Failed](https://github.com/ROCm/TheRock/actions/runs/30665102481/job/91294472528) |
| https://github.com/ROCm/TheRock/commit/d4680c216d0faecc5122cd4a7197368f13f3325a | azure | [Failed](https://github.com/ROCm/TheRock/actions/runs/30687538053/job/91350350837) (driver issue) |
| https://github.com/ROCm/TheRock/commit/3fd408aad89afed286af4d8d00a278f668e47458 | azure | [Passed](https://github.com/ROCm/TheRock/actions/runs/30794114651/job/91653630736) |
| https://github.com/ROCm/TheRock/commit/ad0b692857b8c140867654676db9a08c561bd0b2 | azure | [Passed](https://github.com/ROCm/TheRock/actions/runs/30800288314/job/91670191345) |
| https://github.com/ROCm/TheRock/commit/9cddaa482f3cd1988d36705ce0090bfd75b0127c | CS-RORDMZ | [Failed](https://github.com/ROCm/TheRock/actions/runs/30802799661/job/91678326838) |
| https://github.com/ROCm/TheRock/commit/4d7626337cac97ed87801f3f57224e12fe509860 | CS-RORDMZ | [Failed](https://github.com/ROCm/TheRock/actions/runs/30811359674/job/91711138010) |
| https://github.com/ROCm/TheRock/commit/0888ca35dad1ae9a5f9b46bf56269520fde79cdb | azure | [Passed](https://github.com/ROCm/TheRock/actions/runs/30837625513/job/91798185709) |
| https://github.com/ROCm/TheRock/commit/3eec13b3a0cdbe5816e7c4261318a5a56bf392a9 | CS-RORDMZ | [Failed](https://github.com/ROCm/TheRock/actions/runs/31010753031/job/92356216961) |

--- a/build_tools/github_actions/test_executable_scripts/test_hipthreads_examples.py
+++ b/build_tools/github_actions/test_executable_scripts/test_hipthreads_examples.py
@@ -20,6 +20,7 @@ import logging
 import os
 import platform
 import shlex
+import shutil
 import subprocess
 from pathlib import Path
 
@@ -173,6 +174,13 @@ def configure_and_build(example: dict, gpu_arch: str, environ_vars: dict) -> Pat
     binary = build_dir / "bin" / f"{example['binary']}{EXE_SUFFIX}"
     if not binary.exists():
         raise FileNotFoundError(f"Built binary not found: {binary}")
+
+    if IS_WINDOWS:
+        # Copy artifact DLLs next to the exe so it loads the HIP runtime under test,
+        # not a driver's amdhip64_7.dll shadowing it from System32 (TheRock#7132).
+        for dll in (OUTPUT_ARTIFACTS_PATH / "bin").glob("*.dll"):
+            shutil.copy2(dll, binary.parent)
+
     return binary
 
 

--- a/build_tools/github_actions/test_executable_scripts/test_hipthreads_examples.py
+++ b/build_tools/github_actions/test_executable_scripts/test_hipthreads_examples.py
@@ -15,12 +15,12 @@ copies the source tree, including examples/, into the build dir which the test
 artifact globs as hipthreads/**/*).
 """
 
+import ctypes
 import json
 import logging
 import os
 import platform
 import shlex
-import shutil
 import subprocess
 from pathlib import Path
 
@@ -76,6 +76,19 @@ RUN_TIMEOUT_SECONDS = 1800
 
 IS_WINDOWS = platform.system() == "Windows"
 EXE_SUFFIX = ".exe" if IS_WINDOWS else ""
+
+
+def set_dll_directory(path: str | None) -> None:
+    """Add a DLL search dir ahead of System32.
+
+    PATH can't override a stale amdhip64_7.dll in System32; SetDllDirectoryW can,
+    without copying DLLs into the artifact (TheRock#7132). None restores default.
+    """
+    kernel32 = ctypes.WinDLL("kernel32", use_last_error=True)
+    kernel32.SetDllDirectoryW.argtypes = [ctypes.c_wchar_p]
+    kernel32.SetDllDirectoryW.restype = ctypes.c_bool
+    if not kernel32.SetDllDirectoryW(path):
+        raise ctypes.WinError(ctypes.get_last_error())
 
 
 def load_rocm_version() -> str:
@@ -175,12 +188,6 @@ def configure_and_build(example: dict, gpu_arch: str, environ_vars: dict) -> Pat
     if not binary.exists():
         raise FileNotFoundError(f"Built binary not found: {binary}")
 
-    if IS_WINDOWS:
-        # Copy artifact DLLs next to the exe so it loads the HIP runtime under test,
-        # not a driver's amdhip64_7.dll shadowing it from System32 (TheRock#7132).
-        for dll in (OUTPUT_ARTIFACTS_PATH / "bin").glob("*.dll"):
-            shutil.copy2(dll, binary.parent)
-
     return binary
 
 
@@ -255,6 +262,9 @@ if IS_WINDOWS:
     os.environ["PATH"] = (
         str(OUTPUT_ARTIFACTS_PATH / "bin") + os.pathsep + os.environ.get("PATH", "")
     )
+    # Select the artifact HIP runtime ahead of System32 for every child we spawn
+    # (offload-arch, cmake/clang, the examples). See set_dll_directory.
+    set_dll_directory(str(OUTPUT_ARTIFACTS_PATH / "bin"))
 
 gpu_arch = get_gpu_architecture_portable(OUTPUT_ARTIFACTS_DIR)
 if not gpu_arch:

